astropy/cosmology/tests/test_core.py

```diff
@@ -0,0 +1,555 @@
+
+    def test_init(self, cosmo_cls):
+        # Cosmology accepts any args, kwargs
+        cosmo1 = cosmo_cls(1, 2, 3, 4, 5, a=1, b=2, c=3, d=4, e=5)
+        self._cosmo_test_init_attr(cosmo1)
+
+        # but only "name" and "meta" are used
+        cosmo2 = cosmo_cls(name="test", meta={"m": 1})
+        self._cosmo_test_init_attr(cosmo2)
+        assert cosmo2.name == "test"
+        assert cosmo2.meta["m"] == 1
+
+    # ---------------------------------------------------------------
+
+    def test_is_equivalent(self, cosmo):
+        """
+        Test :meth:`astropy.cosmology.Cosmology.is_equivalent`.
+        This is an abstract method and nothing is equivalent.
+        """
+        # to self
+        assert not cosmo.is_equivalent(cosmo)
+
+        # same class, different instance
+        newclone = cosmo.clone(name="cloned")
+        assert not cosmo.is_equivalent(newclone)
+        assert not newclone.is_equivalent(cosmo)
+
+        # different class
+        assert not cosmo.is_equivalent(2)
+
+
+class TestCosmologySubclass(TestCosmology):
+    """
+    Test subclasses of :class:`astropy.cosmology.Cosmology`.
+    This is broken away from ``TestCosmology``, because |Cosmology| is an ABC
+    and subclasses must override some methods.
+    """
+
+    def setup_class(self):
+        """
+        Setup for testing.
+        Cosmology is abstract, so tests are done on a subclass.
+        """
+        from astropy.cosmology import FLRW  # for simple `__equiv__`
+
+        class SubCosmology(Cosmology):
+            def __equiv__(self, other):  # abstract. Need simple override.
+                return FLRW.__equiv__(self, other)
+
+        self.cls = SubCosmology
+        self.cls_args = (70 * (u.km / u.s / u.Mpc),)  # nothing real
+        self.cls_kwargs = dict(name="test", meta={"a": "b"})
+
+    def teardown_class(self):
+        _COSMOLOGY_CLASSES.pop("TestCosmologySubclass.setup_class.<locals>.SubCosmology", None)
+
+    # ===============================================================
+    # Method & Attribute Tests
+
+    def test_is_equivalent(self, cosmo):
+        """
+        Test :meth:`astropy.cosmology.Cosmology.is_equivalent`, overrode.
+        Normally abstract, but here we set up expectations for all subclasses
+        of |Cosmology|.
+        """
+        # super().test_equivalent(cosmo)  # Not called b/c N/A.
+
+        # to self
+        assert cosmo.is_equivalent(cosmo)
+
+        # same class, different instance
+        newclone = cosmo.clone(name="cloned")
+        assert cosmo.is_equivalent(newclone)
+        assert newclone.is_equivalent(cosmo)
+
+        # different class
+        assert not cosmo.is_equivalent(2)
+
+
+# -----------------------------------------------------------------------------
+
+
+class FlatCosmologyMixinTest:
+    """Test :class:`astropy.cosmology.core.FlatCosmologyMixin`."""
+
+    def test_init(self, cosmo_cls):
+        super().test_init(cosmo_cls)
```

Comment:
Yikes, sometimes pytest really just confuses me: where is `super()` supposed to go in a class that has no superclasses!?

Reply:
Ahh yes. ``pytest`` doesn't actually test this class until it is in mixed into a class starting with ``Test``.
So it will have a superclass then.

Reply:
OK, I missed that it didn't start with `Test`... Resolved...

Reply:
Actually I'm still confused why this is necessary, as it *just* calls `super()`

Reply:
Ah. Thanks. Yes, this one is no longer necessary.
I like to have a test for every changed method on a subclass.
Even if no new tests are needed, having the test just call super tells me the method is different on the subclass, and covered by the test suite. Makes reading the test suite informative about the code. A personal preference. But this mixin no longer has the method… so I’ll remove it here.